### ISSUE TITLE
Silent failures when creating users with SG REST API

### ISSUE THREAD

**Issue by farzadghanei** [floragunncom/search-guard#628]
Hi

We're creating users with internal users REST API, while the API responses with success, a portion of users are not created at all, or can not login with the chosen password. No error/warning, only INFO logs from SG when creating the users.

### Context
 * user creation and login is done by automation software process, no human interaction
 * passwords are alphanumeric, auto generated.
 * for the same user, running the same process would fix the issue, but even the 2nd attempt is not successful all the times, 3rd or 4th attempt succeeds
 * REST API is called in small batches (around 5-10 concurrent requests)
 * requests are load balanced to multiple API servers, ElasticSearch cluster is healthy.
 * Nginx handles requests and does SSL termination, proxies to SG REST API

Here are some Nginx logs (usernames are masked in this post)

```
root@kibana1.internal ~ # grep -F '24/Jan/2019:09:56' /var/log/nginx/kibana.*.log | grep -P 'PUT .+searchguard'

[24/Jan/2019:09:56:19 +0100] "PUT /rest/_searchguard/api/internalusers/********* HTTP/1.1" 200 55 "-" "python-requests/2.19.1"
[24/Jan/2019:09:56:20 +0100] "PUT /rest/_searchguard/api/internalusers/********* HTTP/1.1" 200 59 "-" "python-requests/2.19.1"
[24/Jan/2019:09:56:20 +0100] "PUT /rest/_searchguard/api/internalusers/********* HTTP/1.1" 200 53 "-" "python-requests/2.19.1"
[24/Jan/2019:09:56:21 +0100] "PUT /rest/_searchguard/api/internalusers/********* HTTP/1.1" 200 58 "-" "python-requests/2.19.1"
[24/Jan/2019:09:56:21 +0100] "PUT /rest/_searchguard/api/internalusers/********* HTTP/1.1" 201 65 "-" "python-requests/2.19.1"
[24/Jan/2019:09:56:21 +0100] "PUT /rest/_searchguard/api/internalusers/********* HTTP/1.1" 201 65 "-" "python-requests/2.19.1"
[24/Jan/2019:09:56:34 +0100] "PUT /rest/_searchguard/api/internalusers/********* HTTP/1.1" 200 56 "-" "python-requests/2.19.1"
```

For all the 7 requests above, we have a corresponding `INFO` level log from SG (installation specific data are masked in this post):

```
root@kibana1.internal ~ # grep -F '2019-01-24T09:56' /var/log/elasticsearch/elasticsearch.log | grep -v 'Authentication finally failed for'
[2019-01-24T09:56:18,989][INFO ][c.f.s.c.ComplianceConfig ] Compliance features are disabled. To enable them you need a special license. Please contact support for this.
[2019-01-24T09:56:18,990][INFO ][c.f.s.c.IndexBaseConfigurationRepository] Search Guard License Info: SearchGuardLicense [uid=****************, type=FULL, features=[], issueDate=*****, expiryDate=*****, issuedTo=Byte B.V, issuer=floragunn GmbH, startDate=******, majorVersion=6, clusterName=*, allowedNodeCount=32768, msgs=[], expiresInDays=*****, isExpired=false, valid=true, action=, prodUsage=Yes, one cluster with all commercial features and unlimited nodes per cluster., clusterService=org.elasticsearch.cluster.service.ClusterService@*****, getMsgs()=[], getExpiresInDays()=*****, isExpired()=false, isValid()=true, getAction()=, getProdUsage()=Yes, one cluster with all commercial features and unlimited nodes per cluster.]

(same log lines are repeated for other requests)
```

Both users created with requests above with `201` responses, had issues logging in. I requested for both users again. but the `GET` call to API only returned 1 of the users, the other was not created.

### System info:
 * Debian GNU/Linux 8.10 (jessie)
 * Java: 
    * OpenJDK Runtime Environment (build 1.8.0_171-8u171-b11-1~bpo8+1-b11)
    * OpenJDK 64-Bit Server VM (build 25.171-b11, mixed mode)
 * ES and Kibana: `6.2.3`
 * Kibana SG plugin: `6.2.3-14`
 * SG modules:
    * Audit Logging (enterprise, `6.2.3.32-0`)
    * Document and Field-level security (enterprise, `6.2.3.32-0`)
    * HTTP Basic Authenticator (`6.2.3.23-0`)
    * Internal users authentication (`6.2.3.23-0`)
    * Kibana Multitenancy (enterprise, `6.2.3.32-0`)
    * REST Management API (enterprise, `6.2.3.32-0`)

solutions/hints are much appreciated.

**Comment by floragunncom** [comment]
@jochenkressin can you have a look?

**Comment by jochenkressin** [comment]
Hi Jeroen,

thanks for sharing the script. We have identified the issue and are working on a fix which will hopefully be released this week. To give you some background: When updating the SG configuration index via the REST API, it is first loaded, then the changes are applied, and then it is written back. This leads to race conditions when performing changes in parallel.

**Comment by floragunncom** [comment]
@farzadghanei @JeroenvHeugten if you like you can help us verify that our fix is correct. For that, I can provide a snapshot for ES 6.2.3.

**Comment by farzadghanei** [comment]
@floragunncom yes, we'd be glad to help on this.

**Comment by floragunncom** [comment]
@farzadghanei pls try https://oss.sonatype.org/content/repositories/snapshots/com/floragunn/search-guard-6/6.2-HEAD-restlock-SNAPSHOT/search-guard-6-6.2-HEAD-restlock-20190128.213118-2.zip

**Comment by JeroenvHeugten** [comment]
@floragunncom I tested out the provided snapshot release. It seems you have implemented a global lock on the API. I have to admit I was hoping for another solution, so that at least concurrent requests would succeed, but slower (instead of failing with a hard error). 

> {"status":"INTERNAL_SERVER_ERROR","message":"Error [sg][internalusers]: version conflict, current version [1209] is different than the one provided [1208]"}

**Comment by floragunncom** [comment]
@JeroenvHeugten yes, that's correct, we used the optimistic locking feature. The client can just retry the call if status is not 2xx. So it's consistent with the ES API in general. For the next feature release we think of a global queue but that's not a trivial task because a client can issue requests concurrently to different nodes.

**Comment by floragunncom** [comment]
released with v24.1

**Comment by farzadghanei** [comment]
@floragunncom may we know how we could have this fix on 6.2.x branch?

**Comment by floragunncom** [comment]
Any chance to upgrade to >= 6.4? If not pls contact us directly via email https://search-guard.com/contacts/
